### PR TITLE
Fix conversion of unitless variables to IP in tabular output and added missing conversions

### LINKED ISSUE #5668: EnergyPlus cannot convert unitless variables to IP in tabular output
## Issue overview

Variables with no units (e.g., ZONE PEOPLE OCCUPANT COUNT[]) get converted to a value of zero when `OutputControl:Table:Style` is set to `InchPound`.

They are set to zero in `GetUnitConversion()` because there are no blank units (e.g. `""`) defined in `UnitConv()`. `LookupSItoIP()` doesn't return a non-zero value for `indexUnitConv` [here](https://github.com/NREL/EnergyPlus/blob/fc0a53d6ebe461c2e9a0514df9ea0f97b4b9a42f/src/EnergyPlus/OutputReportTabular.cc#L6403).
### Details
- EnergyPlus, Version 8.5.0-c87e61b44b
### Checklist
- [x] Defect file (e.g., testfiles/PurchAirTables.idf)
- [ ] Ticket added to Pivotal for defect (development team task)
- [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by nealkruis**
## Pull request overview

Fixes #5668.

SI to IP unit conversions were not handled correctly for some units (including unitless quantities) in the tabular output. In these cases the resulting IP value would be zero. This fix
1. explicitly adds several new unit conversions,
2. makes the default conversion multiplier 1.0, and 
3. prints a warning to alert the user when an explicit conversion factor cannot be found.
### Work Checklist
- [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
- [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
  - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
  - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
  - NewFeature: This pull request includes code to add a new feature to EnergyPlus
  - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
  - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog
### Review Checklist

This will not be exhaustively relevant to every PR.
- [x] Code style (parentheses padding, variable names)
- [x] Functional code review (it has to work!)
- [x] If defect, results of running current develop vs this branch should exhibit the fix
- [x] CI status: all green or justified
- [x] Performance: CI Linux results include performance check -- verify this
- [x] Unit Test(s)

### CODE REVIEW

**Comment by nealkruis** [comment]
@JasonGlazer I took a crack at fixing this. You're more familiar with how the conversion process works, so there may be a more efficient way of handling this, or reasons not to fix this the way I did. For example, why do the positions of the brackets bounding the units have to be [greater than or equal to two apart](https://github.com/NREL/EnergyPlus/pull/5669/files#diff-cfc84b456bd550d5729f5275fb755f77L14643)? I couldn't find any exceptions where two neighboring brackets (e.g. `[]`) would break anything.

**Comment by JasonGlazer** [comment]
@nealkruis If I remember correctly, the intention was that zero would represent an unfound unit conversion like a blank. Given this, I would rather see the bracket stuff remain as is (since it only looks when a character exists between the brackets) but change the GetUnitConversion() routine to return a 1.0 when unitConvIndex = 0 (see line 14936 of OutputReportTabular.cc - GitHub won't display that line)

**Comment by JasonGlazer** [comment]
I think getSpecificUnitMultiplier(), getSpecificUnitDivider(), ConvertIPdelta() should be altered too to return the SIvalue when the unitConvIndex is zero. It looks like ConvertIP() already does that.

**Comment by nealkruis** [comment]
@JasonGlazer Are there currently cases where we anticipate "unfound" unit conversions? Wouldn't it be better to reserve the "zero" unit index for unhandled cases and post a warning or an error to the user?

In my opinion, it might be better to explicitly handle the case of unitless values since they are common and expected rather than lumping them in with the exception case.

**Comment by nealkruis** [comment]
Interesting the CI is coming back green. I'd expect big table diffs for most test files with "InchPound" tables.

**Comment by mjwitte** [comment]
Probably because CI only runs design days which leaves many tables as all zeroes.

-------- Original message --------

**Comment by nrel-bot-3** [comment]
@JasonGlazer @nealkruis @lgentile it has been 14 days since this pull request was last updated.

**Comment by mjwitte** [comment]
@nealkruis Is this ready for review?  I don't see a pivotal story for this.  Did you run any local test suites with annual simulations to see the expected diffs?

**Comment by nealkruis** [comment]
I was waiting for a response from @JasonGlazer regarding my proposal to add warnings for "unfound" unit conversions. This is not DOE funded so I didn't add a pivotal story. Not sure what the protocol is in this case. @lgentile?

**Comment by JasonGlazer** [comment]
@nealkruis Sorry, I didn't respond sooner. I don't have an issue with zero being an unsupported unit and providing an error message. I think you just need to be careful with the code since that is a change in approach. As you see, you can't depend too much on the CI for this kind of change, you need to think about doing some test cases locally and making sure everything works as expected.

**Comment by nealkruis** [comment]
@JasonGlazer I came across several other unsupported units when testing locally and added them to the conversion list (I believe these were all previously converted to 0.0). Now the default conversion for unsupported units is 1.0, and a warning will be added to the ERR file when encountered. The only unsupported unit I've found so far that wasn't a 1.0 conversion was "W/C" for a UA value.

I just pushed a change that will force all tables to be converted to IP. I expect BIG diffs since the CI will be comparing SI and IP tables, but for now, I'm just looking for ERR file diffs to see if the warning was triggered. The new warning should tell us if there any other unsupported units found in the test files. There may be other unsupported units that aren't covered in the test files, that we might just have to fix over time.

**Comment by nealkruis** [comment]
@JasonGlazer I'm running a slightly different test through just to be sure. I'll then revert the changes of the last test commit if all comes back good.

Two things I'd like your input on:
1. It's possible there are still unsupported units that won't be uncovered by the integration tests in the CI. I think it would be a lot of work to track down every possible SI unit that could be passed into LookupSItoIP. We might have to address them as users uncover them. Does that seem reasonable? Do you have other ideas?
2.  The warning users will see is shown on [this line](https://github.com/NREL/EnergyPlus/pull/5669/files#diff-cfc84b456bd550d5729f5275fb755f77R14806). Let me know if you think it should read differently.

**Comment by nealkruis** [comment]
The three files with ERR diffs were all the same unit problems related to the Fan:ComponentModel:

```
   ** Warning ** Unable to find a unit conversion from Drive Ratio [-] into IP units
   **   ~~~   ** Applying default conversion factor of 1.0
   **   ~~~   ** Please report to https://github.com/NREL/EnergyPlus/issues
   ** Warning ** Unable to find a unit conversion from Design Belt Output Torque [Nm] into IP units
   **   ~~~   ** Applying default conversion factor of 1.0
   **   ~~~   ** Please report to https://github.com/NREL/EnergyPlus/issues
   ** Warning ** Unable to find a unit conversion from Design Fan Efficiency  [-] into IP units
   **   ~~~   ** Applying default conversion factor of 1.0
   **   ~~~   ** Please report to https://github.com/NREL/EnergyPlus/issues
   ** Warning ** Unable to find a unit conversion from Maximum Belt Efficiency [-] into IP units
   **   ~~~   ** Applying default conversion factor of 1.0
   **   ~~~   ** Please report to https://github.com/NREL/EnergyPlus/issues
   ** Warning ** Unable to find a unit conversion from Design Belt Efficiency [-] into IP units
   **   ~~~   ** Applying default conversion factor of 1.0
   **   ~~~   ** Please report to https://github.com/NREL/EnergyPlus/issues
   ** Warning ** Unable to find a unit conversion from Maximum Motor Efficiency [-] into IP units
   **   ~~~   ** Applying default conversion factor of 1.0
   **   ~~~   ** Please report to https://github.com/NREL/EnergyPlus/issues
   ** Warning ** Unable to find a unit conversion from Design Motor Efficiency [-] into IP units
   **   ~~~   ** Applying default conversion factor of 1.0
   **   ~~~   ** Please report to https://github.com/NREL/EnergyPlus/issues
   ** Warning ** Unable to find a unit conversion from Design VFD Efficiency [-] into IP units
   **   ~~~   ** Applying default conversion factor of 1.0
   **   ~~~   ** Please report to https://github.com/NREL/EnergyPlus/issues
   ** Warning ** Unable to find a unit conversion from Design Combined Efficiency [-] into IP units
   **   ~~~   ** Applying default conversion factor of 1.0
   **   ~~~   ** Please report to https://github.com/NREL/EnergyPlus/issues
```

Does anyone see a problem with changing the `[-]` to `[]` to be consistent with the other unitless outputs? I'll also add a torque conversion to the list.

**Comment by JasonGlazer** [comment]
@nealkruis I think many of the unit conversions are not tested in the integration tests. Maybe we should be adding unit tests for them. The error message seems ok. I am not sure about the suggestion to post the issue to github since we don't do that anywhere else, maybe @mjwitte can chime in on that.

**Comment by nealkruis** [comment]
Part of the problem is we don't know what units are possibly passed to `LookupSItoIP`. If we had a definitive list of SI units, we could define exhaustive unit tests, but I don't think we do (correct?). Integration tests uncovered a few more, but there's no way to know if we're done without a definitive list.

My thought is: A warning is better than defaulting quietly. Hopefully users will never see it, but if they do they can help make the unit list a bit more exhaustive by reporting it. 

I am not sure about the suggestion to post an issue either. There certainly hasn't been a precedent for it. What are the potential downsides?

**Comment by JasonGlazer** [comment]
Well searching for SetupOutputVariable, the units used there are:
[ ]
[%]
[]
[A]
[ach]
[Ah]
[C]
[cd/m2]
[clo]
[deg]
[deltaC]
[hr]
[J/kg]
[J/kg-K]
[J/kgWater]
[J/m2]
[J]
[K/m]
[kg/kg]
[kg/m3]
[kg/s]
[kg]
[kgWater/kgDryAir]
[kgWater/s]
[kmol/s]
[L]
[lum/W]
[lux]
[m/s]
[m]
[m2]
[m3/s]
[m3]
[min]
[Pa]
[ppm]
[rad]
[rev/min]
[s]
[V]
[W/K]
[W/m2]
[W/m2-C]
[W/m2-K]
[W/W]
[W]

Some Excel magic.

File: src/EnergyPlus/OutputReportTabular.cc [hunk 1]
```diff
@@ -14749,7 +14798,16 @@ Label900: ;
...
 		//CALL  ShowWarningError('LookupSItoIP in: ' // TRIM(stringInWithSI) // ' out: ' // TRIM(stringOutWithIP))
 		//IF (foundConv .NE. 0) CALL  ShowWarningError('   Hint ' // TRIM(UnitConv(foundConv)%hint) // IntToStr(foundConv) )
+
 		unitConvIndex = selectedConv;
+
+		// Add warning if units not found. Exclude stirngs without square brackets, or Rating metrics that are inheirently IP (e.g. SEER, IEER, EER, IPLV)
+		if (unitConvIndex == 0 && ! noBrackets && unitSIOnly != "BTU/W-H") {
+			ShowWarningError("Unable to find a unit conversion from " + stringInWithSI + " into IP units");
+			ShowContinueError("Applying default conversion factor of 1.0");
+			ShowContinueError("Please report to https://github.com/NREL/EnergyPlus/issues");
```
Comment (mjwitte) [inline src/EnergyPlus/OutputReportTabular.cc]: @nealkruis @JasonGlazer I would drop the last two lines of the warning.  Can't find a conversion, so it won't be converted.  I don't think it adds anything for the user to say more.

**Comment by mjwitte** [comment]
@nealkruis @JasonGlazer I've often thought that the units should be stored in a separate data field from the variable name and possibly passed as an integer/enum value instead of a string.  The corresponding SI string, IP string, and conversion would all be predefined, and if a new unit is needed, it won't work until these are added to the table.  And the unit type designator could then have multiple flavors for W such as electricW, refrigerationW, etc. for cases where there is more than one IP unit for a given SI unit.  Of course, this would mean modifying _every_ SetUpOutputVariable call, but that may be doable with a smart script.  Probably beyond the scope of this fix, but worth considering as a followup to help standardize this and reduce the string searching.

**Comment by JasonGlazer** [comment]
@mjwitte that would be a good approach, except for the 3855 instances of SetupOutputVariable :-)

**Comment by nealkruis** [comment]
@JasonGlazer we can test all of the brackets found in SetupOutputVariable calls, but there are still other units used in tabular output that aren't accounted for here. Maybe there is a similar way to search those uses as well?

**Comment by nealkruis** [comment]
@mjwitte I 100% agree. Maybe we can define a small task to do this at the planning meeting? I think a regex script should be able to handle 95% of the SetupOutputVariable instances. Most of the work would be in designing the table and refactoring the current functions that interpret units.

File: src/EnergyPlus/OutputReportTabular.cc (hunk 1 above)
Comment (nealkruis) [inline src/EnergyPlus/OutputReportTabular.cc]: I think it's nice to know how the conversion is being handled in the event of an unsupported unit (previously it was multiply by zero).

I suppose the real danger of pointing to GitHub issue tracker is that inexperienced users may misinterpret this as a support service. I'm stepping back on that proposal. I think we would like users to issue a bug report if they come across this, but maybe we let them come to that conclusion.

**Comment by nealkruis** [comment]
There are also EMS output variables where the units are user defined. We definitely don't want to suggest a bug report for these when they're unsupported, but we should probably mention if EnergyPlus doesn't know how to convert the units.

**Comment by mjwitte** [comment]
Added to the planning list.  Forgot about EMS variables.  For now a warning, yes.  But we could consider adding optional fields to EMS:OutputVariable to specify IP Units and the conversion factor.

**Comment by nealkruis** [comment]
Everything is coming back green except one of the unit tests on the Windows machine. It works fine on the other operating systems and on my local Windows machine. @Myoldmopar any thoughts as to what's happening here?

**Comment by nealkruis** [comment]
@Myoldmopar a similar problem seems to be showing up on [develop](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-7e24f8abc0c396da442795ecae964d3981c5e2fd-Win64-Windows-7-VisualStudio-12.html#test) as well.

**Comment by mjwitte** [comment]
@nealkruis These random unit test failures have been showing up for a while, can't open some file, often a different unit test.  If you delete results for that platform so it reruns the tests, it will likely come back clean, but that's not necessary.

**Comment by nealkruis** [comment]
@mjwitte @JasonGlazer I think this is ready for review and merging. Do one of you want to take a look? Let me know if you have any questions.

**Comment by mjwitte** [comment]
@nealkruis Please add a brief summary in the top comment of this PR that explains (from user perspective) what was changed/fixed.

**Comment by mjwitte** [comment]
But also keep the reference to the original issue, of course.

**Comment by JasonGlazer** [comment]
@nealkruis I think "$" and "$/UNIT ENERGY" should be `"~~$~~"` and `"~~$~~/UNIT ENERGY"` so that the monetary unit substitution works, see InsertCurrencySymbol(). 

Also the unit W/C should probably just use W/K which already exists in the list of units.  

Why is line 14835 including: "unitSIOnly != "BTU/W-H" is that being used as an SI unit somewhere?

Now that unfound units are zeros, ConvertIPDelta() needs to return the SI value if the index is zero like in ConvertIP().

Should getSpecificUnitMultiplier() and getSpecificUnitDivider() return a 1.0 if the unit is not found?

**Comment by nealkruis** [comment]
@JasonGlazer 
1. These are hard-set to "$" in the [LEED predefined report](https://github.com/NREL/EnergyPlus/blob/d60482c9f60f218c0a2f4786be22f5e1dcc2cd32/src/EnergyPlus/OutputReportPredefined.cc#L1987). Should I change them to `"~~$~~"`?
2. W/C is ubiquitous throughout the code. Are you suggesting that we change them all to W/K? or that we simply reference the same conversion factor. Now that the units are handled, maybe we leave it as is?
3. The "BTU/W-H" comes from AHRI rating metrics (SEER, EER, IEER) which are defined in I-P units and thus exist in the SI tables as I-P units. Maybe we actually create a 1-to-1 conversion for these?
4. I'll make the two consistent.
5. I assume so. I'll also add a warning here if there is an unhandled unit conversion. There are relatively few instances of these functions, and they all use common unit conversions. I don't think this will ever trigger.

**Comment by JasonGlazer** [comment]
@nealkruis 
1. The $ should probably be cleaned up but maybe it should be a separate issue.
2. The W/C vs W/K should probably be cleaned up sometime also but it also seems outside of the scope at this point.
3. You should add a 1-to-1 conversion for  "BTU/W-H" for now so there is no exception on that line of code.
4. Thanks.
5. Thanks.

**Comment by nealkruis** [comment]
@JasonGlazer that should do it. Feel free to merge if you're happy and it comes back green.

**Comment by JasonGlazer** [comment]
@nealkruis The code all looks fine. Thanks for making those additional changes. Have you double checked all the (non-1.0) unit conversion factors or should I?

**Comment by nealkruis** [comment]
I checked them pretty closely when adding them, but it wouldn't hurt for another pair of eyes to do a pass through.

**Comment by JasonGlazer** [comment]
I checked the conversions and they look fine. I noticed that 109 and 110 were repeats of 26 and 34 respectively. I used [NIST SP811 ](http://www.nist.gov/pml/pubs/sp811/)for a few of them which is a handy resource.

**Pull request merged**